src/agentscope/parsers/regex_tagged_content_parser.py

```diff
@@ -0,0 +1,169 @@
+        r"</\1?>",
+        format_instruction: Optional[str] = None,
+        try_parse_json: bool = True,
+        required_keys: Optional[List[str]] = None,
+        keys_to_memory: Union[str, bool, Sequence[str]] = True,
+        keys_to_content: Union[str, bool, Sequence[str]] = True,
+        keys_to_metadata: Union[str, bool, Sequence[str]] = False,
+    ) -> None:
+        """Initialize the regex tagged content parser.
+
+        Args:
+            tagged_content_pattern (`Optional[str]`, defaults to `None`):
+                The regex pattern to extract tagged content. The pattern should
+                contain two named groups: `name` and `content`. The `name`
+                group is used as the key of the tagged content, and the
+                `content` group is used as the value.
+            format_instruction (`Optional[str]`, defaults to `None`):
+                The instruction for the format of the tagged content, which
+                will be attached to the end of the prompt messages to remind
+                the LLM to follow the format.
+            try_parse_json (`bool`, defaults to `True`):
+                Whether to try to parse the tagged content as JSON. Note
+                the parsing function won't raise exceptions.
+            required_keys (`Optional[List[str]]`, defaults to `None`):
+                The keys that are required in the tagged content.
+            keys_to_memory (`Union[str, bool, Sequence[str]]`,
+            defaults to `True`):
+                The keys to save to memory.
+            keys_to_content (`Union[str, bool, Sequence[str]]`,
+            defaults to `True`):
+                The keys to save to content.
+            keys_to_metadata (`Union[str, bool, Sequence[str]]`,
+            defaults to `False`):
+                The key or keys to be filtered in `to_metadata` method. If
+                it's
+                - `False`, `None` will be returned in the `to_metadata` method
+                - `str`, the corresponding value will be returned
+                - `List[str]`, a filtered dictionary will be returned
+                - `True`, the whole dictionary will be returned
+        """
+
+        DictFilterMixin.__init__(
+            self,
+            keys_to_memory=keys_to_memory,
+            keys_to_content=keys_to_content,
+            keys_to_metadata=keys_to_metadata,
+        )
+
+        assert (
+            "<name>" in tagged_content_pattern
+        ), "The tagged content pattern should contain a named group 'name'."
+        assert (
+            "<content>" in tagged_content_pattern
+        ), "The tagged content pattern should contain a named group 'content'."
+
+        self.tagged_content_pattern = tagged_content_pattern
+        self._format_instruction = format_instruction
+        self.try_parse_json = try_parse_json
+        self.required_keys = required_keys or []
+
+    @property
+    def format_instruction(self) -> str:
```

Comment:
Not an important issue, but should we have a built in format instruction template? Or at least let the user know what their own format instruction should be link?

Reply:
The key problem in regex tagged content parser is that we cannot know the required keys in advanced. So here we can provide a general format instruction here.